packages/botonic-plugin-nlu/src/utils.js

```diff
@@ -0,0 +1,121 @@
+import fs from 'fs'
+import path from 'path'
+import mkdirp from 'mkdirp'
+import { NLU_DATA_FILENAME, MODELS_DIRNAME } from './constants'
+function escapeRegExp(str) {
+  return str.replace(/([.*+?^=!:${}()|\[\]\/\\])/g, '\\$1')
+}
+
+export function replaceAll(str, find, replace) {
+  return str.replace(new RegExp(escapeRegExp(find), 'g'), replace)
+}
+
+function shuffle(obj1, obj2) {
+  // shuffle two related arrays preserving indexes
+  var index = obj1.length
+  var rnd, tmp1, tmp2
+
+  while (index) {
+    rnd = Math.floor(Math.random() * index)
+    index -= 1
+    tmp1 = obj1[index]
+    tmp2 = obj2[index]
+    obj1[index] = obj1[rnd]
+    obj2[index] = obj2[rnd]
+    obj1[rnd] = tmp1
+    obj2[rnd] = tmp2
+  }
+}
+
+export function clone(src) {
+  return Object.assign([], src)
+}
+
+export function readDir(dirPath) {
+  return fs.readdirSync(dirPath)
+}
+
+export function readFile(filePath, fileEncoding = 'utf-8') {
+  return fs.readFileSync(filePath, fileEncoding)
+}
+
+function getIntentName(fileName) {
+  try {
+    return fileName.match(/(.*).txt/)[1]
+  } catch (e) {
+    console.log(
+      `${fileName} is not a valid. File must be of type IntentName.txt`
+    )
+  }
+}
+
+export function loadIntentsData({
+  intentsPath,
+  fileEncoding = 'utf-8',
+  shuffleData = true
+}) {
+  let samples = []
+  let labels = []
+  let intentsDict = {}
+  let intentsFiles = readDir(intentsPath)
+  for (let i = 0, len = intentsFiles.length; i < len; i++) {
+    intentsDict[i] = getIntentName(intentsFiles[i])
+    let file = readFile(path.join(intentsPath, intentsFiles[i])).split('\n')
+    for (let l = 0, len = file.length; l < len; l++) {
```

Comment:
Using "for (let sentence in sentences)" instead avoids having so many intermediate variables

Reply:
Once a time I read that the for loop is the one that gives a better performance compared with map, forEach, etc. Maybe I've done some premature optimization here, but really don't know how this could affect when working with large datasets. I think that should be ideal that the methods involved with making predictions when the bot receives new inputs must be as fast as possible, whereas the rest of methods can be rewritten with more readable methods as you suggest.

Reply:
https://github.com/dg92/Performance-Analysis-JS